File: Samples/RadialMenu/UE4/Source/RadialMenu/RadialPanel.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// NoesisGUI - http://www.noesisengine.com
// Copyright (c) 2013 Noesis Technologies S.L. All Rights Reserved.
////////////////////////////////////////////////////////////////////////////////////////////////////


#include "RadialPanel.h"

#include <NsCore/ReflectionImplement.h>
#include <NsGui/UIElementData.h>
#include <NsGui/FrameworkPropertyMetadata.h>
#include <NsGui/Brush.h>
#include <NsGui/Pen.h>
#include <NsGui/StreamGeometry.h>
#include <NsGui/StreamGeometryContext.h>
#include <NsGui/FreezableCollection.h>
#include <NsGui/UIElementCollection.h>
#include <NsGui/DrawingContext.h>


using namespace RadialMenu;


////////////////////////////////////////////////////////////////////////////////////////////////////
RadialPanel::RadialPanel()
{
}

////////////////////////////////////////////////////////////////////////////////////////////////////
RadialPanel::~RadialPanel()
{
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialPanel::GetAngle() const
{
    return GetValue<float>(AngleProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialPanel::SetAngle(float value)
{
    SetValue<float>(AngleProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialPanel::GetAngleOffset() const
{
    return GetValue<float>(AngleOffsetProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialPanel::SetAngleOffset(float value)
{
    SetValue<float>(AngleOffsetProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
bool RadialPanel::GetIsClockwise() const
{
    return GetValue<bool>(IsClockwiseProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialPanel::SetIsClockwise(bool value)
{
    SetValue<bool>(IsClockwiseProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialPanel::GetCenterX(const UIElement* child)
{
    NS_CHECK(child != nullptr, "Child is null");
    return child->GetValue<float>(CenterXProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialPanel::SetCenterX(UIElement* child, float value)
{
    NS_CHECK(child != nullptr, "Child is null");
    child->SetValue<float>(CenterXProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialPanel::GetCenterY(const UIElement* child)
{
    NS_CHECK(child != nullptr, "Child is null");
    return child->GetValue<float>(CenterYProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialPanel::SetCenterY(UIElement* child, float value)
{
    NS_CHECK(child != nullptr, "Child is null");
    child->SetValue<float>(CenterYProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialPanel::GetAngleStart(const UIElement* child)
{
    NS_CHECK(child != nullptr, "Child is null");
    return child->GetValue<float>(AngleStartProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialPanel::SetAngleStart(UIElement* child, float value)
{
    NS_CHECK(child != nullptr, "Child is null");
    child->SetValue<float>(AngleStartProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float RadialPanel::GetAngleDelta(const UIElement* child)
{
    NS_CHECK(child != nullptr, "Child is null");
    return child->GetValue<float>(AngleDeltaProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void RadialPanel::SetAngleDelta(UIElement* child, float value)
{
    NS_CHECK(child != nullptr, "Child is null");
    child->SetValue<float>(AngleDeltaProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
Noesis::Size RadialPanel::MeasureOverride(const Noesis::Size& availableSize)
{
    Noesis::Size desiredSize;

    Noesis::UIElementCollection* children = GetChildren();
    uint32_t numChildren = children->Count();

    for (uint32_t i = 0; i < numChildren; ++i)
    {
        Noesis::UIElement* child = children->Get(i);

        child->Measure(availableSize);

        const Noesis::Size& childSize = child->GetDesiredSize();
        if (desiredSize.width < childSize.width) desiredSize.width = childSize.width;
        if (desiredSize.height < childSize.height) desiredSize.height = childSize.height;
    }

    return desiredSize;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
Noesis::Size RadialPanel::ArrangeOverride(const Noesis::Size& finalSize)
{
    float diameter = Noesis::Min(finalSize.width, finalSize.height);
    float radius = diameter * 0.5f;

    Noesis::Point center(radius, radius);
    Noesis::Size renderSize(diameter, diameter);
    Noesis::Rect childRect(renderSize);

    Noesis::UIElementCollection* children = GetChildren();
    uint32_t numChildren = children->Count();

    float childAngleStart = GetAngleOffset();
    float childAngleDelta = GetAngle() / numChildren;

    for (uint32_t i = 0; i < numChildren; ++i)
    {
        Noesis::UIElement* child = children->Get(i);

        child->Arrange(childRect);

        SetCenterX(child, center.x);
        SetCenterY(child, center.y);
        SetAngleStart(child, childAngleStart);
        SetAngleDelta(child, childAngleDelta);

        childAngleStart += childAngleDelta;
    }

    return renderSize;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
NS_BEGIN_COLD_REGION

NS_IMPLEMENT_REFLECTION(RadialPanel, "RadialMenu.RadialPanel")
{
    Noesis::UIElementData* data = NsMeta<Noesis::UIElementData>(Noesis::TypeOf<SelfClass>());

    data->RegisterProperty<float>(AngleProperty, "Angle",
        Noesis::FrameworkPropertyMetadata::Create(360.0f,
            Noesis::FrameworkPropertyMetadataOptions_AffectsArrange));
    data->RegisterProperty<float>(AngleOffsetProperty, "AngleOffset",
        Noesis::FrameworkPropertyMetadata::Create(0.0f,
            Noesis::FrameworkPropertyMetadataOptions_AffectsArrange));
    data->RegisterProperty<bool>(IsClockwiseProperty, "IsClockwise",
        Noesis::FrameworkPropertyMetadata::Create(true,
            Noesis::FrameworkPropertyMetadataOptions_AffectsArrange));

    data->RegisterProperty<float>(CenterXProperty, "CenterX",
        Noesis::PropertyMetadata::Create(0.0f));
    data->RegisterProperty<float>(CenterYProperty, "CenterY",
        Noesis::PropertyMetadata::Create(0.0f));
    data->RegisterProperty<float>(AngleStartProperty, "AngleStart",
        Noesis::PropertyMetadata::Create(0.0f));
    data->RegisterProperty<float>(AngleDeltaProperty, "AngleDelta",
        Noesis::PropertyMetadata::Create(0.0f));
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const Noesis::DependencyProperty* RadialPanel::AngleProperty;
const Noesis::DependencyProperty* RadialPanel::AngleOffsetProperty;
const Noesis::DependencyProperty* RadialPanel::IsClockwiseProperty;
const Noesis::DependencyProperty* RadialPanel::CenterXProperty;
const Noesis::DependencyProperty* RadialPanel::CenterYProperty;
const Noesis::DependencyProperty* RadialPanel::AngleStartProperty;
const Noesis::DependencyProperty* RadialPanel::AngleDeltaProperty;

File: Samples/RadialMenu/UE4/Assets/RadialMenu/RadialMenu/RadialPanel.cs
using System;
using System.Windows;
using System.Windows.Controls;

namespace RadialMenu
{
    /// <summary>
    /// Panel to display items in a radial/arc.
    /// </summary>
    public class RadialPanel : Panel
    {
        #region Dependency Properties
        /// <summary>
        /// Gets or sets the arc angle covered by the children. A value between 0 and 360 degrees
        /// </summary>
        public double Angle
        {
            get { return (double)GetValue(AngleProperty); }
            set { SetValue(AngleProperty, value); }
        }

        public static readonly DependencyProperty AngleProperty = DependencyProperty.Register(
            "Angle", typeof(double), typeof(RadialPanel), new FrameworkPropertyMetadata(360.0,
                FrameworkPropertyMetadataOptions.AffectsArrange));

        /// <summary>
        /// Gets or sets the angle offset where children start to be positioned
        /// </summary>
        public double AngleOffset
        {
            get { return (double)GetValue(AngleOffsetProperty); }
            set { SetValue(AngleOffsetProperty, value); }
        }

        public static readonly DependencyProperty AngleOffsetProperty = DependencyProperty.Register(
            "AngleOffset", typeof(double), typeof(RadialPanel), new FrameworkPropertyMetadata(0.0,
                FrameworkPropertyMetadataOptions.AffectsArrange));

        /// <summary>
        /// Gets or sets the arc angle covered by the children. A value between 0 and 360 degrees
        /// </summary>
        public bool IsClockwise
        {
            get { return (bool)GetValue(IsClockwiseProperty); }
            set { SetValue(IsClockwiseProperty, value); }
        }

        public static readonly DependencyProperty IsClockwiseProperty = DependencyProperty.Register(
            "IsClockwise", typeof(bool), typeof(RadialPanel), new FrameworkPropertyMetadata(true,
                FrameworkPropertyMetadataOptions.AffectsArrange));
        #endregion

        #region Attached Properties
        public static double GetCenterX(UIElement obj) { return (double)obj.GetValue(CenterXProperty); }
        public static void SetCenterX(UIElement obj, double value) { obj.SetValue(CenterXProperty, value); }

        public static readonly DependencyProperty CenterXProperty = DependencyProperty.RegisterAttached(
            "CenterX", typeof(double), typeof(RadialPanel), new PropertyMetadata(0.0));

        public static double GetCenterY(UIElement obj) { return (double)obj.GetValue(CenterYProperty); }
        public static void SetCenterY(UIElement obj, double value) { obj.SetValue(CenterYProperty, value); }

        public static readonly DependencyProperty CenterYProperty = DependencyProperty.RegisterAttached(
            "CenterY", typeof(double), typeof(RadialPanel), new PropertyMetadata(0.0));

        public static double GetAngleStart(UIElement obj) { return (double)obj.GetValue(AngleStartProperty); }
        public static void SetAngleStart(UIElement obj, double value) { obj.SetValue(AngleStartProperty, value); }

        public static readonly DependencyProperty AngleStartProperty = DependencyProperty.RegisterAttached(
            "AngleStart", typeof(double), typeof(RadialPanel), new PropertyMetadata(0.0));

        public static double GetAngleDelta(UIElement obj) { return (double)obj.GetValue(AngleDeltaProperty); }
        public static void SetAngleDelta(UIElement obj, double value) { obj.SetValue(AngleDeltaProperty, value); }

        public static readonly DependencyProperty AngleDeltaProperty = DependencyProperty.RegisterAttached(
            "AngleDelta", typeof(double), typeof(RadialPanel), new PropertyMetadata(0.0));
        #endregion

        protected override Size MeasureOverride(Size availableSize)
        {
            Size desiredSize = new Size(0, 0);
            foreach (UIElement child in Children)
            {
                child.Measure(availableSize);

                if (desiredSize.Width < child.DesiredSize.Width) desiredSize.Width = child.DesiredSize.Width;
                if (desiredSize.Height < child.DesiredSize.Height) desiredSize.Height = child.DesiredSize.Height;
            }

            return desiredSize;
        }

        protected override Size ArrangeOverride(Size finalSize)
        {
            var diameter = Math.Min(finalSize.Width, finalSize.Height);
            var radius = diameter * 0.5;

            Point center = new Point(radius, radius);
            Size renderSize = new Size(diameter, diameter);
            Rect childRect = new Rect(renderSize);

            int childCount = Children.Count;
            double childAngleStart = AngleOffset;
            double childAngleDelta = Angle / childCount;

            foreach (UIElement child in Children)
            {
                child.Arrange(childRect);

                SetCenterX(child, center.X);
                SetCenterY(child, center.Y);
                SetAngleStart(child, childAngleStart);
                SetAngleDelta(child, childAngleDelta);

                childAngleStart += childAngleDelta;
            }

            return renderSize;
        }
    }
}
